contracts/estate/EstateRegistry.sol

```diff
@@ -0,0 +1,236 @@
+pragma solidity ^0.4.22;
+
+import "zeppelin-solidity/contracts/math/SafeMath.sol";
+import "zeppelin-solidity/contracts/token/ERC721/ERC721Token.sol";
+import "zeppelin-solidity/contracts/ownership/Ownable.sol";
+
+import "../metadata/MetadataHolderBase.sol";
+
+
+contract PingableDAR {
+  function ping() public;
+  function ownerOf(uint256 tokenId) public returns (address);
+  function safeTransferFrom(address, address, uint256) public;
+}
+
+
+// @nico TODO: Standalone Storage.
+// @nico TODO: Document functions.
+contract EstateRegistry is ERC721Token, Ownable, MetadataHolderBase {
+  using SafeMath for uint256;
+
+  PingableDAR public dar;
+
+  // From Estate to list of owned tokenIds (LANDs)
+  mapping(uint256 => uint256[]) public estateTokenIds;
+
+  // From tokenId (LAND) to its owner estateId of owned tokenIds
+  mapping(uint256 => uint256) public tokenIdEstate;
+
+  // From Estate to mapping of tokenId to index on the array above
+  mapping(uint256 => mapping(uint256 => uint256)) public estateTokenIndex;
+
+  // Metadata of the Estate
+  mapping(uint256 => string) internal estateData;
+
+  // Operators of the Estate
+  mapping (uint256 => address) public updateOperator;
+
+  constructor(
+    string _name,
+    string _symbol,
+    address _dar
+  )
+    ERC721Token(_name, _symbol)
+    Ownable()
+    public
+  {
+    require(_dar != 0);
+    dar = PingableDAR(_dar);
+  }
+
+  modifier onlyDAR() {
+    require(msg.sender == address(dar));
+    _;
+  }
+
+  modifier onlyAuthorized(uint256 estateId) {
+    require(_isAuthorized(msg.sender, estateId), "Unauthorized user");
+    _;
+  }
+
+  modifier onlyUpdateAuthorized(uint256 estateId) {
+    require(_isUpdateAuthorized(msg.sender, estateId), "Unauthorized user");
+    _;
+  }
+
+  function mintNext(address to) public returns (uint256) {
+    uint256 tokenId = _getNewTokenId();
+    _mint(to, tokenId);
+    return tokenId;
+  }
+
+  // onERC721Received: Count
+  function onERC721Received(
+    address /* oldOwner */,
+    uint256 tokenId,
+    bytes estateTokenIdBytes
+  )
+    external
+    onlyDAR
+    returns (bytes4)
+  {
+    if(estateTokenIdBytes.length > 0) {
+      uint256 estateId = uint256(_bytesToBytes32(estateTokenIdBytes));
+      _pushTokenId(estateId, tokenId);
+    }
+    return bytes4(0xf0b9e5ba);
+  }
+
+  function ammendReceived(uint256 estateId, uint256 tokenId) external {
+    _pushTokenId(estateId, tokenId);
```

Comment:
Event here?

Reply:
I added and later removed an `AmmendReceived` event. I'll add it again!

Reply:
What do you think of adding a modifier `onlyOwner(estateId)` to avoid third parties to emit event and modified the storage of the contract?